### PR TITLE
SeqAlgorithm bug fixes

### PR DESCRIPTION
**Pull request by lucas-FP**
Created SeqWrapper abstract class. Sequence algorithms should implement the `get_inner_signature` function in order to allow grammars to be built correctly.

Changed pipeline graph builder to traverse algorithms in the `pool` of `Algorithm`s and `SeqAlgorithm`s instead of the given registry.

### CODE REVIEW

File: autogoal/kb/_algorithm.py [hunk 1]
```diff
@@ -491,7 +511,7 @@ def run(self, a: type_1, b: type_2, ...) -> type_n:
         guaranteed_types = node.output_types
 
         # Here are all the algorithms that could be added new at this point in the graph
-        for algorithm in registry:
+        for algorithm in pool:
```
Comment (apiad) [inline autogoal/kb/_algorithm.py:514]: This looks like a very silly mistake on our side, honestly :sweat: so I'm wondering why are there no tests that cover this!

File: autogoal/kb/_algorithm.py [hunk 2]
```diff
@@ -319,6 +332,12 @@ def repr_method(self):
     def getattr_method(self, attr):
         return getattr(self.inner, attr)
 
+    @classmethod
```
Comment (apiad) [inline autogoal/kb/_algorithm.py:335]: Can we add some kind of test (either as docstring or as a proper test in `/tests/core`) that illustrates why this change is necessary? I know it's after the fact, but the idea is if we reverse these changes, is there some fairly simple use case that breaks? Maybe it's not trivial to find one, but I'd like to at least make the thought experiment of trying to come up with one.

File: autogoal/kb/_algorithm.py [hunk 3]
```diff
@@ -339,8 +358,9 @@ def body(ns):
         ns["input_types"] = input_types_method
         ns["input_args"] = input_args_method
         ns["output_type"] = output_types_method
+        ns["get_inner_signature"] = get_inner_signature_method
 
-    return types.new_class(name=name, bases=(Algorithm,), exec_body=body)
+    return types.new_class(name=name, bases=(Algorithm, SeqWrapper), exec_body=body)
```
Comment (apiad) [inline autogoal/kb/_algorithm.py:363]: Why do we need explicit inheritance here, if `get_inner_signature` is solved via introspection? Do you see any advantage in explicitly introducing a new class to the API (given that the introduced method is abstract, so we're not inheriting any functionality) vs just keeping it as an implicit convention?

**Review by apiad — COMMENTED** [review]
Hey, @lucas-FP thanks for the PR! I left some comments just to flesh out the motivation for this change. Let me know what you think.

**Comment by apiad** [comment]
Damn! The stupid memory test failed, I have no idea why, but it's probably a fault on the test itself. Let's not focus on that for now.

File: autogoal/kb/_algorithm.py (hunk 3 above)
Comment (lucas-FP) [reply autogoal/kb/_algorithm.py:363]: No advantage indeed. I probably tried solving it in a different way with this class but then just left it there afterwards.

File: autogoal/kb/_algorithm.py (hunk 2 above)
Comment (lucas-FP) [reply autogoal/kb/_algorithm.py:335]: Did not think of or found any use case that is broken by this change. I've searched for other uses of the `inspect.signature` method, but all of them were being used with the `run` method and not the `__init__`.  There is one in the demo though ([this line](https://github.com/autogoal/autogoal/blob/136cb8c9a41ca59fafad56ee8a22a3c888a99b74/autogoal/contrib/streamlit/demo.py#L320)) but i think this is not going to break it.

Comment (lucas-FP) [reply autogoal/kb/_algorithm.py:323]: Do you think the docstring test is enough @apiad ?

**Pull request merged**